content/blog/2021-02-17-cml-runner-prerelease.md

```diff
@@ -0,0 +1,165 @@
+    runs-on: [self-hosted, cml-runner]
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-python@v2
+        with:
+          python-version: '3.x'
+      - name: 'Train my model'
+        run: |
+          pip install -r requirements.txt
+          python train.py
+```
+
+If you use CML functions in the `train-model` step, you can go even further and
+get a closed loop- sending model training results from the EC2 instance to your
+pull request or merge request! For example, if we expand the `train-model` step
+to incorporate functions like `cml-publish` and `cml-send-comment`:
+
+```yaml
+train-model:
+  needs: deploy-runner
+  runs-on: [self-hosted, cml-runner]
+  container: docker://dvcorg/cml-py3
+  steps:
+    - uses: actions/checkout@v2
+    - name: 'Train a model'
+      env:
+        repo_token: ${{ secrets.PERSONAL_ACCESS_TOKEN }}
+      run: |
+        pip install -r requirements.txt
+        python train.py
+
+        echo "## Report from your EC2 Instance" > report.md
+        cat metrics.txt >> report.md
+        cml-publish "confusion_matrix.png" --md >> report.md
+        cml-send-comment report.md
+```
+
+You'll get a pull request that looks something like this:
+
+![](/uploads/images/2021-02-17/sample_pr.png)
+
+All the code to replicate this example is up on a
+[brand new demo repository](https://github.com/iterative/cml-runner-base-case).
+
+### Our favorite details
+
+The new `cml-runner` function lets you turn on instances, including GPU,
+high-memory and sport instances, and kick off a new workflow using the hardware
+and environment of your choice- and of course, it'll turn _off_ those instances
+after a configurable timeout!
+
+Another highlight: you can use whatever Docker container you'd like on your
+instance. In the above example, we use our
+[custom CML Docker container](https://github.com/iterative/cml/blob/master/docker/Dockerfile)
+(because we like it!) - but you certainly don't have to! Whatever image you
+choose, we highly recommend containerizing your environment for ultimate
+reproducibility and security with CML.
+
+You can also use the new `cml-runner` function to set up a local self-hosted
+runner. On the local machine, you'll install CML as a package and then run:
+
+```bash
+$ cml-runner \
+    --repo $your_project_repository_url \
+    --token=$personal_access_token \
+    --labels tf \
+    --idle-timeout 180
+```
+
+Now your machine will be listening for workflows from your project repository.
+
+## A New GitHub Action
+
+One more thing: you might've noticed in our example workflow above that there's
+a [new CML GitHub Action](https://github.com/iterative/cml-setup)! The new
+Action helps you setup CML, giving you one more way to mix and match the CML
+suite of functions with your preferred environment.
+
+In addition to the CML library, including functions like `cml-send-comment`,
+`cml-publish` and `cml-runner`, using the `setup-cml` Action also gives you DVC
```

Comment:
setup-cml does not give you dvc out of the box. For that you have setup-dvc

Reply:
I thought that was funny- I got this detail from[ the README on `setup-cml`](https://github.com/iterative/setup-cml):

> The main difference between this and the CML container is that the latest can work also as a self-hosted runner with GPU capabilites and a cloud provisioner. Additionally comes packed with DVC. As said this action only installs the functions available in CML to generate comments, graphs and checks.

So I will suggest that revision in my docs cleanup there!

Reply:
https://github.com/iterative/setup-cml/issues/12